tests/robustness/model/replay.go

```diff
@@ -19,82 +19,86 @@ import (
-	if revision < 1 {
-		return EtcdState{}, fmt.Errorf("invalid revision: %d", revision)
-	}
-	if r.cachedState == nil || r.cachedState.Revision > revision {
-		r.reset()
-	}
+func (r *EtcdReplay) StateForRevision(revision int64) EtcdState {
+	return r.revisionToEtcdState[revision]
+}
 
-	for r.eventHistoryIndex < len(r.eventHistory) && r.cachedState.Revision < revision {
-		nextRequest, nextRevision, nextIndex := r.next()
-		newState, _ := r.cachedState.Step(nextRequest)
-		if newState.Revision != nextRevision {
-			return EtcdState{}, fmt.Errorf("model returned different revision than one present in event history, model: %d, event: %d", newState.Revision, nextRevision)
+func (r *EtcdReplay) EventsForWatch(watch WatchRequest) (events []PersistedEvent) {
+	for _, e := range r.Events {
+		if e.Revision < watch.Revision || !e.Match(watch) {
+			continue
 		}
-		r.cachedState = &newState
-		r.eventHistoryIndex = nextIndex
+		events = append(events, e)
 	}
-	if r.eventHistoryIndex > len(r.eventHistory) && r.cachedState.Revision < revision {
-		return EtcdState{}, fmt.Errorf("requested revision higher then available in even history, requested: %d, model: %d", revision, r.cachedState.Revision)
-	}
-	return *r.cachedState, nil
+	return events
 }
 
-func (r *EtcdReplay) reset() {
-	state := freshEtcdState()
-	r.cachedState = &state
-	r.eventHistoryIndex = 0
-}
-
-func (r *EtcdReplay) next() (request EtcdRequest, revision int64, index int) {
-	revision = r.eventHistory[r.eventHistoryIndex].Revision
-	index = r.eventHistoryIndex
-	operations := []EtcdOperation{}
-	for index < len(r.eventHistory) && r.eventHistory[index].Revision == revision {
-		event := r.eventHistory[index]
-		switch event.Type {
-		case PutOperation:
-			operations = append(operations, EtcdOperation{
-				Type: event.Type,
-				Put:  PutOptions{Key: event.Key, Value: event.Value},
-			})
-		case DeleteOperation:
-			operations = append(operations, EtcdOperation{
-				Type:   event.Type,
-				Delete: DeleteOptions{Key: event.Key},
-			})
+func toWatchEvents(prevState *EtcdState, request EtcdRequest, response MaybeEtcdResponse) (events []PersistedEvent) {
+	if request.Type == Txn && response.Error == "" {
+		var ops []EtcdOperation
+		if response.Txn.Failure {
+			ops = request.Txn.OperationsOnFailure
+		} else {
+			ops = request.Txn.OperationsOnSuccess
+		}
+		for _, op := range ops {
+			switch op.Type {
+			case RangeOperation:
+			case DeleteOperation:
+				e := PersistedEvent{
+					Event: Event{
+						Type: op.Type,
+						Key:  op.Delete.Key,
+					},
+					Revision: response.Revision,
+				}
+				if _, ok := prevState.KeyValues[op.Delete.Key]; ok {
+					events = append(events, e)
+				}
+			case PutOperation:
+				e := PersistedEvent{
+					Event: Event{
+						Type:  op.Type,
+						Key:   op.Put.Key,
+						Value: op.Put.Value,
+					},
+					Revision: response.Revision,
+				}
+				if _, ok := prevState.KeyValues[op.Put.Key]; !ok {
+					e.IsCreate = true
```

Comment:
I had a question about how we are detecting an event to be a create event: iiuc, we look at prevState and if the key is not present in its `KeyValues`, we say it is a create event. Can there be a situation where a delete operation takes place and deletes the key?

Based on this
https://github.com/etcd-io/etcd/blob/dd4e35a58597e66b74a84a5a99ad42a7b7723415/tests/robustness/model/deterministic.go#L140

I understand that as we `Step` through a delete request, that will delete the key from the `KeyValues` map, given that behaviour, is it possible that we misattribute an event as a create event?

Reply:
Not sure I understand the question.

Reply:
Nvm, I misunderstood it a little bit, this looks good!